flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java

```diff
@@ -0,0 +1,283 @@
+
+		StreamGraphUtils.validateTransformationUid(context.getStreamGraph(), transformation);
+
+		int writerId = addWriter(transformation, context);
+		int committerId = addCommitter(
+				writerId,
+				transformation,
+				new StreamingCommitterOperatorFactory<>(transformation.getSink()),
+				context);
+		addGlobalCommitter(
+				committerId >= 0 ? committerId : writerId,
+				transformation,
+				new GlobalStreamingCommitterOperatorFactory<>(transformation.getSink()),
+				context);
+
+		return Collections.emptyList();
+	}
+
+	/**
+	 * Add a sink writer node to the stream graph.
+	 *
+	 * @param sinkTransformation The transformation that the writer belongs to
+	 *
+	 * @return The stream node id of the writer
+	 */
+	private int addWriter(
+			SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT> sinkTransformation,
+			Context context) {
+		final boolean hasState = sinkTransformation
+				.getSink()
+				.getWriterStateSerializer()
+				.isPresent();
+		final TypeInformation<CommT> outTypeInfo = TypeExtractor.createTypeInfo(
+				Sink.class,
+				sinkTransformation.getSink().getClass(),
+				1,
+				null,
+				null);
+		@SuppressWarnings("unchecked")
+		final Transformation<InputT> input = (Transformation<InputT>) sinkTransformation
+				.getInputs()
+				.get(0);
+		final TypeInformation<InputT> inputTypeInfo = input.getOutputType();
+
+		final int parallelism = getParallelism(sinkTransformation, context);
+
+		final StreamOperatorFactory<CommT> writer =
+				hasState ? new StatefulWriterOperatorFactory<>(sinkTransformation.getSink()) : new StatelessWriterOperatorFactory<>(
+						sinkTransformation.getSink());
+
+		final int writerId = addOperatorToStreamGraph(
+				writer, input.getId(),
+				inputTypeInfo,
+				outTypeInfo,
+				"Sink Writer:",
+				parallelism,
+				sinkTransformation.getMaxParallelism(),
+				sinkTransformation,
+				context);
+
+		StreamGraphUtils.configureResourceProperties(
+				context.getStreamGraph(),
+				writerId,
+				sinkTransformation);
+		return writerId;
+	}
+
+	/**
+	 * Try to add a sink committer to the stream graph.
+	 *
+	 * @param inputId The committer's input stream node id
+	 * @param sinkTransformation The transformation that the committer belongs to
+	 * @param committerFactory The committer operator's factory
+	 *
+	 * @return The stream node id of the committer or -1 if the sink topology does not include a committer.
+	 */
+	private int addCommitter(
+			int inputId,
+			SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT> sinkTransformation,
+			OneInputStreamOperatorFactory<CommT, CommT> committerFactory,
+			Context context) {
+
+		if (!sinkTransformation.getSink().getCommittableSerializer().isPresent()) {
+			return -1;
+		}
+
+		final TypeInformation<CommT> committableTypeInfo = TypeExtractor.createTypeInfo(
+				Sink.class,
+				sinkTransformation.getSink().getClass(),
+				1,
+				null,
+				null);
+
+		final int parallelism = getParallelism(sinkTransformation, context);
```

Comment:
Are we sure that in both `BATCH` and `STREAMING` the parallelism of the `Committter` will be equal to `parallelism`? Why not passing it as an argument and figuring the actual value in the `translateForStreaming` or `translateForBatch`?

Reply:
I agree this would need some discussion.